Azure Arc AKS on HCI (#388)
microsoft/azure_arc: docs/azure_arc_jumpstart/azure_arc_k8s/aks_stack_hci/aks_hci_powershell/_index.md

---
type: docs
title: "AKS on Azure Stack HCI PowerShell"
linkTitle: "AKS on Azure Stack HCI PowerShell"
weight: 1
description: >
---

> ## Deploy AKS cluster on Azure Stack HCI and connect it to Azure Arc using PoweSh

Reviewer: ```suggestion
## Deploy AKS cluster on Azure Stack HCI and connect it to Azure Arc using PoweShell
```

> The following README will guide you on how to use the provided PowerShell script to deploy an [Azure Kubernetes Service (AKS)](https://docs.microsoft.com/en-us/azure/aks/intro-kubernetes) cluster on [Azure Stack HCI](https://docs.microsoft.com/en-us/azure-stack/hci/overview) and connected it as an Azure Arc cluster resource.

Reviewer: ```suggestion
The following README will guide you on how to use the provided PowerShell script to deploy an [Azure Kubernetes Service (AKS)](https://docs.microsoft.com/en-us/azure/aks/intro-kubernetes) cluster on [Azure Stack HCI](https://docs.microsoft.com/en-us/azure-stack/hci/overview) and connected it as an Azure Arc enabled Kubernetes cluster resource.
```

Azure Kubernetes Service on Azure Stack HCI is an implementation of AKS on-premises using hyperconverged infrastructure operating system that is delivered as an Azure service.  

>   > **Note: Azure Kubernetes Service is now in preview on Azure Stack HCI and on Windows Server 2019 Datacenter.**

Reviewer: ```suggestion
  > **Note: Azure Kubernetes Service is currently in preview on Azure Stack HCI and on Windows Server 2019 Datacenter.**
```

  > **Note: Currently, Azure Arc enabled Kubernetes is in [public preview](https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/).**

> This guide assumes you already have an existing 2-4 node Azure Stack HCI cluster and will leverage PowerShell to automate the AKS creation and onboarding process of the cluster to Azure Arc. 

Reviewer: ```suggestion
This guide assumes you already have an existing Azure Stack HCI cluster (with 2 to 4 nodes) and will leverage PowerShell to automate the AKS creation and onboarding process of the cluster to Azure Arc. 
```

Reviewer: 1. Is 4 the supported maximum number of nodes?
2. IMO, the language here should be more decisive in the sense of why we are assuming this and not providing the instructions on deploying HCI.

    ServicePrincipalNames : {XXXXXXXXXXXXXXXXXXXXXXXXXXXX, http://AzureArcK8s}
    ApplicationId         : XXXXXXXXXXXXXXXXXXXXXXXXXXXX
    ObjectType            : ServicePrincipal
    DisplayName           : AzureArcK8s
    Id                    : XXXXXXXXXXXXXXXXXXXXXXXXXXXX
    Type                  :
    ```

    To expose the generated password use this code to export the secret: 

    ```poweshell
    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($sp.Secret)
    $UnsecureSecret = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
    ```

    > **Note: It is optional but highly recommended to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/powershell/module/az.resources/new-azadserviceprincipal?view=azps-5.4.0)**

* Enable your subscription with the two resource providers for Azure Arc enabled Kubernetes. Registration is an asynchronous process, and registration may take approximately 10 minutes.

  ```powershell
  Register-AzResourceProvider -ProviderNamespace Microsoft.Kubernetes
  Register-AzResourceProvider -ProviderNamespace Microsoft.KubernetesConfiguration
  ```

> * Since AKS on Azure Stack HCI is in public preview you will need to download the required software for evaluation from the [registration page](https://aka.ms/AKS-HCI-Evaluate). Complete the preview registration form and safe the zip file to the "Downloads" folder.

Reviewer: ```suggestion
* Since AKS on Azure Stack HCI is in preview you will need to download the required software for evaluation from the [registration page](https://aka.ms/AKS-HCI-Evaluate). Complete the preview registration form and save the zip file to the "Downloads" folder.
```

> * Perform a clean install of the AksHci PowerShell module. To install the AksHci Powershell module found in the zip file you just downloaded follow [this guide.](https://docs.microsoft.com/en-us/azure-stack/aks-hci/setup-powershell#step-1-download-and-install-the-akshci-powershell-module)

Reviewer: ```suggestion
* Perform a clean installation of the _AksHci_ PowerShell module. To install the AksHci PowerShell module found in the zip file you just downloaded, follow [this guide.](https://docs.microsoft.com/en-us/azure-stack/aks-hci/setup-powershell#step-1-download-and-install-the-akshci-powershell-module)
```

Reviewer: I think we should have this as part of the scenario

 
## Deployment

> * Before running the PowerShell Script to deploy your AKS cluster, it is important to verify each physical node of your Azure Stack HCI to validate all of the requirements are satisfied. To do that, open PowerShell as an administrator and run:

Reviewer: something about that sentence does not make sense

  ```powershell
  Initialize-AksHciNode
  ```

> * Now that all nodes are ready, you will deploy an AKS cluster to your Azure Stack HCI using this [PowerShell script](https://github.com/microsoft/azure_arc/blob/main/azure_arc_k8s_jumpstart/aks_stack_hci/powershell/aks_hci_deploy.ps1). Note, that the script will deploy a simple DHCP based cluster on your Azure Stack HCI, there are additional optional parameters that you could use to customize the deployment to your own environment as described [here] (https://docs.microsoft.com/en-us/azure-stack/aks-hci/setup-powershell). 
> 

Reviewer: ```suggestion
* Now that all nodes are ready, you will deploy an AKS cluster to your Azure Stack HCI using this [PowerShell script](https://github.com/microsoft/azure_arc/blob/main/azure_arc_k8s_jumpstart/aks_stack_hci/powershell/aks_hci_deploy.ps1). 
Note that the script will deploy a simple DHCP-based cluster on your Azure Stack HCI and there are additional optional parameters that you could use to customize the deployment to your own environment as described [here] (https://docs.microsoft.com/en-us/azure-stack/aks-hci/setup-powershell). 

```

  ```powershell
  .\aks_hci_deploy.ps1
  ```
![Screenshot showing the script input](./02.png)

* The script will prompt to provide a set of parameters to configure your cluster: 
    - **imageDir:** path to the directory where AKS on Azure Stack HCI will store its VHD images, provide a shared path or SMB for multinode
    - **cloudConfigLocation:** path to the directory where the cloud agent will store its configuration, provide a shared path or SMB for multinode
    - **vnetName:** the name of the virtual switch to connect the virtual machines to. If you already have an external switch on the host, you should pass the name of the switch here. 
    - **clusterName:** a name for your AKS cluster, must be lowercase. 
    - **controlPlaneNodeCount:** number of nodes for your control plane, should be an odd number 1, 3 or 5.
    - **linuxNodeCount:** number of Linux node VMs for your cluster, if you do not need Linux nodes input 0.
    - **windowsNodeCount:** number of Windows node VMs for your cluster, if you do not need Windows nodes input 0.

>     > **Note: the script may take upto 30 minutes to run**

Reviewer: ```suggestion
    > **Note: the script may take up to 30 minutes to complete**
```

* Once the script has completed its run, you will see an output as follows: 

![Screenshot showing the script output](./03.png)

* To be able to connect to the Azure Kubernetes Service cluster run the command: 

  ```powershell
  Get-AksHciCredential -clusterName <cluster_name>
  ```
![Screenshot showing the output retrieving credentials](./04.png)

> 

Reviewer: Consider adding more screenshots of the HCI cluster in the Windows Admin Center to provide the reader how an end result should look like

Author: I did not use Windows Admin Center and don't even have it installed as it was another procedure, not sure if it looks the same in a real life deployment compared to nested, I'll try

## Connecting to Azure Arc

> * Now that you have a running AKS cluster, edit the environment variables section in the included [az_connect_aks](https://github.com/microsoft/azure_arc/blob/main/azure_arc_k8s_jumpstart/aks_stack_hci/powershell/aks_hci_connect.ps1) shell script.
> 

Reviewer: ```suggestion
* Now that you have a running AKS cluster, edit the environment variables section in the included [az_connect_aks](https://github.com/microsoft/azure_arc/blob/main/azure_arc_k8s_jumpstart/aks_stack_hci/powershell/aks_hci_connect.ps1) PowerShell script.

```

![Screenshot showing environment variables aks_hci_connect.ps1](./05.png)

> * After editing the environment variables in the aks_hci_connect PowerShell script to match your parameters, save the file and then run it using the .\aks_hci_connect.ps1 command.

Reviewer: ```suggestion
* After editing the environment variables in the aks_hci_connect PowerShell script to match your parameters, save the file and then run it using the *.\aks_hci_connect.ps1* command.
```

Reviewer: Provide and example of the command with real parameters

> * Once the script run has finished, the AKS cluster on HCI will be projected as a new Azure Arc cluster resource.

Reviewer: ```suggestion
* Once the script run has finished, the AKS cluster on HCI will be projected as a new Azure Arc enabled Kubernetes cluster resource.
```

> ![Screenshot showing environment variables aks_hci_connect.ps1](./06.png)

Reviewer: Before showing the cluster resource, have a screenshot showing the cluster as part of the resource group

## Delete the deployment

The most straightforward way is to delete the Azure Arc cluster resource via the Azure Portal, just select the cluster and delete it.

![Screenshot showing how to delete Azure Arc enabled Kubernetes resource](./07.png)

If you want to delete the AKS cluster on HCI run the below command.

```powershell
Uninstall-AksHci -Force

> ```

Reviewer: Add screenshot
